File: mycuration/src/main/java/com/phicdy/mycuration/presentation/presenter/RegisterFilterPresenter.java
package com.phicdy.mycuration.presentation.presenter;

import android.support.annotation.NonNull;
import android.view.MenuItem;

import com.phicdy.mycuration.R;
import com.phicdy.mycuration.data.db.DatabaseAdapter;
import com.phicdy.mycuration.filter.Filter;
import com.phicdy.mycuration.data.rss.Feed;
import com.phicdy.mycuration.presentation.view.RegisterFilterView;

import java.util.ArrayList;

public class RegisterFilterPresenter implements Presenter {

    private RegisterFilterView view;
    private final DatabaseAdapter dbAdapter;
    private ArrayList<Feed> selectedFeedList = new ArrayList<>();

    private int editFilterId = NEW_FILTER_ID;
    private boolean isEdit = false;
    private static final int NEW_FILTER_ID = -1;

    public RegisterFilterPresenter(DatabaseAdapter adapter, int editFilterId) {
        this.dbAdapter = adapter;
        this.editFilterId = editFilterId;
        isEdit = editFilterId != NEW_FILTER_ID;
    }

    public void setView(@NonNull RegisterFilterView view) {
        this.view = view;
        if (isEdit) {
            Filter editFilter = dbAdapter.getFilterById(editFilterId);
            if (editFilter != null) {
                view.setFilterTitle(editFilter.getTitle());
                view.setFilterUrl(editFilter.getUrl());
                view.setFilterKeyword(editFilter.getKeyword());
                setSelectedFeedList(editFilter.feeds());
            }
        }
    }

    public void setSelectedFeedList(ArrayList<Feed> list) {
        this.selectedFeedList = list;
        setTargetRssTitle(selectedFeedList);
    }

    public ArrayList<Feed> selectedFeedList() {
        return selectedFeedList;
    }

    private void setTargetRssTitle(ArrayList<Feed> feeds) {
        if (feeds == null || feeds.size() == 0) {
            view.resetFilterTargetRss();
        } else if (feeds.size() == 1) {
            view.setFilterTargetRss(feeds.get(0).getTitle());
        } else {
            view.setMultipleFilterTargetRss();
        }
    }

    @Override
    public void create() {
    }

    @Override
    public void resume() {
    }

    @Override
    public void pause() {

    }

    private void addMenuClicked() {
        String keywordText = view.filterKeyword();
        String filterUrlText = view.filterUrl();
        String titleText = view.filterTitle();

        //Check title and etKeyword or filter URL has the text
        if (titleText.equals("")) {
            view.handleEmptyTitle();
        } else if ((keywordText.equals("")) && (filterUrlText.equals(""))) {
            view.handleEmptyCondition();
        } else if (keywordText.equals("%") || filterUrlText.equals("%")) {
            view.handlePercentOnly();
        } else {
            boolean result;
            if (isEdit) {
                result = dbAdapter.updateFilter(editFilterId, titleText, keywordText, filterUrlText, selectedFeedList);
                view.trackEdit();
            } else {
                // Add new filter
                result = dbAdapter.saveNewFilter(titleText, selectedFeedList, keywordText, filterUrlText);
                view.trackRegister();
            }
            if (result) {
                view.showSaveSuccessToast();
            }else {
                view.showSaveErrorToast();
            }
            view.finish();
        }
    }

    public void optionItemClicked(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.add_filter:
                addMenuClicked();
                break;
            default:
        }
    }
}

File: mycuration/src/main/java/com/phicdy/mycuration/presentation/presenter/RegisterFilterPresenter.kt
package com.phicdy.mycuration.presentation.presenter

import android.view.MenuItem
import com.phicdy.mycuration.R
import com.phicdy.mycuration.data.db.DatabaseAdapter
import com.phicdy.mycuration.data.repository.FilterRepository
import com.phicdy.mycuration.data.rss.Feed
import com.phicdy.mycuration.presentation.view.RegisterFilterView
import kotlinx.coroutines.coroutineScope
import java.util.ArrayList

class RegisterFilterPresenter(private val view: RegisterFilterView,
                              private val dbAdapter: DatabaseAdapter,
                              private val filterRepository: FilterRepository,
                              private val editFilterId: Int = NEW_FILTER_ID) {

    private var selectedFeedList = ArrayList<Feed>()

    private var isEdit = false

    fun setSelectedFeedList(list: ArrayList<Feed>) {
        selectedFeedList = list
        setTargetRssTitle(selectedFeedList)
    }

    fun selectedFeedList(): ArrayList<Feed> {
        return selectedFeedList
    }

    private fun setTargetRssTitle(feeds: ArrayList<Feed>) {
        when {
            feeds.size == 0 -> view.resetFilterTargetRss()
            feeds.size == 1 -> view.setFilterTargetRss(feeds[0].title)
            else -> view.setMultipleFilterTargetRss()
        }
    }

    suspend fun create() = coroutineScope {
        isEdit = editFilterId != NEW_FILTER_ID
        if (isEdit) {
            filterRepository.getFilterById(editFilterId)?.let {
                view.setFilterTitle(it.title)
                view.setFilterUrl(it.url)
                view.setFilterKeyword(it.keyword)
                setSelectedFeedList(it.feeds)
            }
        }
    }

    private fun addMenuClicked() {
        val keywordText = view.filterKeyword()
        val filterUrlText = view.filterUrl()
        val titleText = view.filterTitle()

        //Check title and etKeyword or filter URL has the text
        when {
            titleText.isBlank() -> view.handleEmptyTitle()
            keywordText.isBlank() && filterUrlText.isBlank() -> view.handleEmptyCondition()
            keywordText == "%" || filterUrlText == "%" -> view.handlePercentOnly()
            else -> {
                val result: Boolean
                if (isEdit) {
                    result = dbAdapter.updateFilter(editFilterId, titleText, keywordText, filterUrlText, selectedFeedList)
                    view.trackEdit()
                } else {
                    // Add new filter
                    result = dbAdapter.saveNewFilter(titleText, selectedFeedList, keywordText, filterUrlText)
                    view.trackRegister()
                }
                if (result) {
                    view.showSaveSuccessToast()
                } else {
                    view.showSaveErrorToast()
                }
                view.finish()
            }
        }
    }

    fun optionItemClicked(item: MenuItem) {
        when (item.itemId) {
            R.id.add_filter -> addMenuClicked()
        }
    }

    companion object {
        private const val NEW_FILTER_ID = -1
    }
}
